eventqueue/processor.go

```diff
@@ -0,0 +1,229 @@
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package eventqueue
+
+import (
+	"errors"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	kclock "k8s.io/utils/clock"
+)
+
+// ErrProcessorStopped is returned when the processor is not running.
+var ErrProcessorStopped = errors.New("processor is stopped")
+
+// Processor manages the queue of items and processes them at the correct time.
+type Processor[T queueable] struct {
+	executeFn          func(r T)
+	queue              *queue[T]
+	clock              kclock.Clock
+	queueLock          sync.Mutex
+	processorRunningCh chan struct{}
+	stopCh             chan struct{}
+	resetCh            chan struct{}
+	stopped            atomic.Bool
+}
+
+// NewProcessor returns a new Processor object.
+// executeFn is the callback invoked when the item is to be executed; this will be invoked in a background goroutine.
+func NewProcessor[T queueable](executeFn func(r T), clock kclock.Clock) *Processor[T] {
+	return &Processor[T]{
+		executeFn:          executeFn,
+		queue:              newQueue[T](),
+		processorRunningCh: make(chan struct{}, 1),
+		stopCh:             make(chan struct{}),
+		resetCh:            make(chan struct{}, 1),
+		clock:              clock,
+	}
+}
+
+// Enqueue adds a new item to the queue.
+// If a item with the same ID already exists, it'll be replaced.
+func (p *Processor[T]) Enqueue(r T) error {
+	if p.stopped.Load() {
+		return ErrProcessorStopped
+	}
+
+	// Insert or replace the item in the queue
+	// If the item added or replaced is the first one in the queue, we need to know that
+	p.queueLock.Lock()
+	peek, ok := p.queue.Peek()
+	isFirst := (ok && peek.Key() == r.Key()) // This is going to be true if the item being replaced is the first one in the queue
+	p.queue.Insert(r, true)
+	peek, _ = p.queue.Peek()         // No need to check for "ok" here because we know this will return an item
+	isFirst = isFirst || (peek == r) // This is also going to be true if the item just added landed at the front of the queue
+	p.process(isFirst)
+	p.queueLock.Unlock()
+
+	return nil
+}
+
+// Dequeue removes a item from the queue.
+func (p *Processor[T]) Dequeue(key string) error {
+	if p.stopped.Load() {
+		return ErrProcessorStopped
+	}
+
+	// We need to check if this is the next item in the queue, as that requires stopping the processor
+	p.queueLock.Lock()
+	peek, ok := p.queue.Peek()
+	p.queue.Remove(key)
+	if ok && peek.Key() == key {
+		// If the item was the first one in the queue, restart the processor
+		p.process(true)
+	}
+	p.queueLock.Unlock()
+
+	return nil
+}
+
+// Stop the processor.
+func (p *Processor[T]) Close() error {
+	if !p.stopped.CompareAndSwap(false, true) {
+		// Already stopped
```

Comment:
We should move the close channel to this if block, and always wait for the processor to finishing running when calling Close.

Reply:
To the block where there's currently the "Already stopped" comment? But then we may close the channel twice which would cause a panic?

Reply:
Yes, sorry- we will need to remove the `!` operator on the if statement

Reply:
Ok should be good now.

Reply:
I think this is still relevant- I think `Close()` should always block until the processor has stopped running for all callers.